src/aspire/source/__init__.py

```diff
@@ -370,6 +375,120 @@ def whiten(self, noise_filter):
+        around the origin (supposed to correspond to the molecule) with the mean
+        of the noise, and making sure that the mean of the molecule is larger.
+        From the implementation level, we modify the `ImageSource` in-place by
+        appending a `Multiple` filter to the generation pipeline.
+
+        :param batch_size: Batch size of images to query.
+        :return: On return, the `ImageSource` object has been modified in place.
+        """
+
+        logger.info('Apply contrast inversion on source object')
+        L = self.L
+        grid = grid_2d(L, shifted=True)
+        # Get mask indices of signal and noise samples assuming molecule
+        signal_mask = (grid['r'] < 0.5)
+        noise_mask = (grid['r'] > 0.8)
+
+        # Calculate mean values in batch_size
+        signal_mean = 0.0
+        noise_mean = 0.0
+
+        for i in range(0, self.n, batch_size):
+            images = self.images(i, batch_size).asnumpy()
+            signal = images * np.expand_dims(signal_mask, 2)
+            noise = images * np.expand_dims(noise_mask, 2)
+            signal_mean += np.sum(signal)
+            noise_mean += np.sum(noise)
+        signal_denominator = self.n * np.sum(signal_mask)
+        noise_denominator = self.n * np.sum(noise_mask)
+        signal_mean /= signal_denominator
+        noise_mean /= noise_denominator
+
+        if signal_mean < noise_mean:
+            logger.info('Need to invert contrast')
+            scale_factor = -1.0 * np.ones(self.n)
+        else:
+            logger.info('No need to invert contrast')
+            scale_factor = 1.0 * np.ones(self.n)
+
+        logger.info('Adding Scaling Xform to end of generation pipeline')
+        self.generation_pipeline.add_xform(Multiply(scale_factor))
+
+        # Invalidate images
+        self._im = None
+
+    def normalize_background(self, bg_radius=1.0, do_ramp=True,
+                             batch_size=512):
+        """
+        Normalize the images by the noise background
+
+        This is done by shifting the image density by the mean value of background
+        and scaling the image density by the standard deviation of background.
+        From the implementation level, we modify the `ImageSource` in-place by
+        appending the `Add` and `Multiple` filters to the generation pipeline.
+
+        :param bg_radius: Radius cutoff to be considered as background (in image size)
+        :param do_ramp: When it is `True`, fit a ramping background to the data
+            and subtract. Namely perform normalization based on values from each image.
+            Otherwise, a constant background level from all images is used.
+        :param batch_size: Batch size of images to query.
+        :return: On return, the `ImageSource` object has been modified in place.
+        """
+
+        logger.info(f'Normalize background on source object with radius size of {bg_radius}')
+
+        if do_ramp:
```

Comment:
This is very confusing. Both branches do the mean and variance normalization, but in different ways (one through this `do_ramping`, which is also confusingly named, since it also does the mean and variance normalization) and one by precalculating the mean and variance and applying them using the `Add` and `Multiply` `Xform`s.

What we had decided to do was to have some equivalent of the Matlab `im_normalize_bg` that is called as part of the pipeline through a `LambdaXform`. That function would essentially be the same as this `do_ramping`, except it would have an argument `do_ramp` that would trigger the ramp subtraction part. This way there's no duplication and everything calculated at pipeline execution time, regardless of how `do_ramp` is set.

Reply:
The do_ramp was renamed to `normalize_bg` and moved to `Image/__init__.py` as a function to be included the `Image` class. New flag was added for the `do_ramp` option. All changes were included in the new commit  90e117d.